src/main/java/org/tools4j/hoverraft/command/InMemoryCommandLog.java

```diff
@@ -32,76 +33,88 @@
+        throw new IllegalArgumentException("Index " + index + " is not in [0, " + (entries.size() - 1) + "]");
     }
 
     @Override
-    public synchronized void append(final LogEntry logEntry) {
-        entries.add(clone(logEntry));
+    public void append(final int term, final Command command) {
+        final LogEntry logEntry = directFactory.logEntry();
+        logEntry.logKey().term(term);
+        logEntry.command()
+                .writeBufferOrNull()
+                .putBytes(0, command.readBufferOrNull(), command.offset(), command.byteLength());
+        synchronized (this) {
+            final long index = entries.size();
+            logEntry.logKey().index(index);//TODO this is redundant, do we need this?
+            entries.add(logEntry);
+            maxCommandIndexBySourceId.put(command.commandKey().sourceId(), command.commandKey().commandIndex());
+        }
     }
 
 
     @Override
     public synchronized void truncateIncluding(final long index) {
-        if (index >= entries.size()) {
-            throw new IllegalArgumentException("Truncate index " + index + " must be less than the size " + entries.size());
-        }
-        for (long idx = lastIndex(); idx >= index; idx--) {
-            entries.remove((int)idx);
+        for (int idx = entries.size() - 1; idx >= index; idx--) {
+            final LogEntry removed = entries.remove(idx);
+            maxCommandIndexBySourceId.remove(removed.command().commandKey().sourceId());
         }
     }
 
     @Override
-    public long lastIndex() {
-        return read(entries.size() - 1).logKey().index();
+    public synchronized int lastTerm() {
+        return readTerm(entries.size() - 1);
     }
 
     @Override
-    public void lastKeyTo(final LogKey logKey) {
-        readKeyTo(entries.size() - 1, logKey);
+    public synchronized void lastKeyTo(final LogKey target) {
+        readKeyTo(entries.size() - 1, target);
     }
 
     @Override
-    public int lastKeyCompareTo(final LogKey logKey) {
-        return read(entries.size() - 1).logKey().compareTo(logKey);
-    }
-
-    private LogEntry clone(final LogEntry e) {
-        final LogEntry clone = directFactory.logEntry();
-        clone.writeBufferOrNull().putBytes(0, e.readBufferOrNull(), e.offset(), e.byteLength());
-        return clone;
+    public synchronized boolean contains(final CommandKey commandKey) {
+        final int sourceId = commandKey.sourceId();
+        long maxCommandIndex = maxCommandIndexBySourceId.get(sourceId);
+        if (maxCommandIndex < 0) {
+            for (int idx = entries.size() - 1; idx >= 0; idx--) {
+                final CommandKey ckey = entries.get(idx).command().commandKey();
+                if (ckey.sourceId() == sourceId) {
+                    maxCommandIndex = ckey.commandIndex();
+                    maxCommandIndexBySourceId.put(sourceId, maxCommandIndex);
+                    break;
+                }
+            }
```

Comment:
I see that you are rebuilding the command index.
Is it a right place in contains() method? Shouldn't it be part of truncate method?